README.md
![License: GPL v2](https://img.shields.io/badge/License-GPL%20v2-blue.svg)

[![Build Status](https://travis-ci.com/udovicic/magento-static-test.svg?branch=master)](https://travis-ci.com/udovicic/magento-static-test)

## What is this?

A Docker image that utilizes a phpcs to evaluate coding standards on code.

Based on:
* [Magento coding standard 31](https://github.com/magento/magento-coding-standard)
* [Magento EQP 2.0.1](https://github.com/magento/marketplace-eqp)
* [Variable Analysis](https://github.com/sirbrillig/phpcs-variable-analysis)

## How to use this image

Basic usage.

```bash
docker run --rm \
    --volume /local/path:/project \
    udovicic/magentost[:tag] [<options>]
```

For example, to check `src` directory against the lMEQP2 coding standard.

```bash
docker run --rm \
    --volume `pwd`:/project \
    udovicic/magentost:latest --standard=mcga --extensions=js,php,phtml src
```

A personal recommendation is to add following to your set of aliases:

```bash
alias mcga='docker run --rm --volume `pwd`:/project udovicic/magentost:latest --standard=mcga --extensions=js,php,phtml'
alias mcga-fix='docker run --rm --entrypoint "phpcbf" --volume `pwd`:/project udovicic/magentost:latest --standard=mcga --extensions=js,php,phtml'
```

In that case, you can simply use this as:

```bash
$ mcga <path_to_source>
```

And to fix what can be automaticall done:
```bash
$ mcga-fix <path_to_source>
```


## Volumes

* `/project`: Your PHP project directory.


CHANGELOG.md
# 31.1

* Update VariableAnalysis package

# 31.0

* Update Magento coding standard to version 31

# 27.1

* Removal of `Missing short description` test

# 27.0

* Update Magento coding standard to version 27
* Remove autotagging to `latest`
* Change in versioning format

# 3.2.0

* Update Magento coding standards to version 6
* Update code sniffer to latest version
* Add escaper and secureRenderer to the lsit of allowed undefined variables
* Set image to be built for PHP 7.4

# 3.1.6

* Ban Magento copyright notice in custom code
* declare(strict_types) must have one empty line between <?php and itself

# 3.1.5

* Update of base Magento coding standard to version 3

# 3.1.4

* Tagging change, meta version

# 3.1.3

* Tagging change, meta version

# 3.1.2

* Comment for function parameter no longer require period at the end
* Comment for function parameter can start with lower case letter

# 3.1.1

* Comment is no longer required for @throws DocBlock
* Comment of @throws no longer require period at the end
* Comment of @throws can start with lower case letter

# 3.1.0

* Introduced new standard, **mcga**, which is supposed to combine _MEQP2_ and _Magento2_ standards + a lot of our own to fix common mistakes. This one is supposed to be frequently expanded
* _mcga_ checks if DocBlock properly describes input and output parameters of the functions, including @throw
* _mcga_ checks if there are unused or undefined variables in code
* _mcga_ forces new line between statement and DocBlock
* _mcga_ requires DocBlock of constant to add more value if it already exists
* Removing vendor from git, refactoring build process
* Bumping base Docker image to PHP 7.3 (previously 5.6)
* License change
* README.md revamp
* Testing on multiple PHP versions


tests/test.php
<?php
/**
 * Tools for static testing Magento code
 * Copyright (C) 2019 Stjepan Udovičić
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

class Test {
    public function doNothingTest() {
        //
    }
}

?>


toolset/custom-standards/mcga/Tests/PHP/DeclareOnNewLineUnitTest.php
<?php
/**
 * Tools for static testing Magento code
 * Copyright (C) 2019 Stjepan Udovičić
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

namespace mcga\Tests\PHP;

use PHP_CodeSniffer\Tests\Standards\AbstractSniffUnitTest;

/**
 * Class DeclareOnNewLineUnitTest
 */
class DeclareOnNewLineUnitTest extends AbstractSniffUnitTest
{
    /**
     * @inheritdoc
     */
    public function getErrorList()
    {
        return [];
    }

    /**
     * @inheritdoc
     */
    public function getWarningList($testFile = '')
    {
        switch ($testFile) {
            case 'DeclareOnNewLineUnitTest.1.inc':
                return [
                    1 => 1
                ];
            case 'DeclareOnNewLineUnitTest.2.inc':
                return [
                    2 => 1
                ];
            default:
                return [];
        }
    }
}


toolset/custom-standards/mcga/Tests/Commenting/PHPDocNewLineUnitTest.php
<?php
/**
 * Tools for static testing Magento code
 * Copyright (C) 2019 Stjepan Udovičić
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

namespace mcga\Tests\Commenting;

use PHP_CodeSniffer\Tests\Standards\AbstractSniffUnitTest;

/**
 * Class PHPDocNewLine
 */
class PHPDocNewLineUnitTest extends AbstractSniffUnitTest
{
    /**
     * @inheritdoc
     */
    public function getErrorList()
    {
        return [];
    }

    /**
     * @inheritdoc
     */
    public function getWarningList($testFile = '')
    {
        return [
            9 => 1
        ];
    }
}


toolset/custom-standards/mcga/Sniffs/PHP/DeclareOnNewLineSniff.php
<?php
/**
 * Tools for static testing Magento code
 * Copyright (C) 2019 Stjepan Udovičić
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

namespace mcga\Sniffs\PHP;

use PHP_CodeSniffer\Sniffs\Sniff;
use PHP_CodeSniffer\Files\File;

/**
 * Enforces declare strict types to be on line of its own
 */
class DeclareOnNewLineSniff implements Sniff
{
    /**
     * @inheritDoc
     */
    public function register()
    {
        return [
            T_DECLARE
        ];
    }

    /**
     * @inheritDoc
     */
    public function process(File $phpcsFile, $stackPtr)
    {
        $tokens = $phpcsFile->getTokens();

        if ($tokens[$stackPtr + 2]['code'] !== T_STRING || $tokens[$stackPtr + 2]['content'] !== 'strict_types' ) {
            return;
        }

        if ($stackPtr >= 2 && $tokens[$stackPtr - 2]['code'] === T_OPEN_TAG) {
            return;
        }

        $phpcsFile->addWarning(
            'declare(strict_types) must have one empty line between <?php and itself',
            $stackPtr,
            'declareNewLineMissing'
        );
    }
}


toolset/custom-standards/mcga/Sniffs/Commenting/PHPDocNewLineSniff.php
<?php
/**
 * Tools for static testing Magento code
 * Copyright (C) 2019 Stjepan Udovičić
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

namespace mcga\Sniffs\Commenting;

use PHP_CodeSniffer\Sniffs\Sniff;
use PHP_CodeSniffer\Files\File;

/**
 * Detects PHPDoc formatting for constants.
 */
class PHPDocNewLineSniff implements Sniff
{
    /**
     * @inheritDoc
     */
    public function register()
    {
        return [
            T_DOC_COMMENT_OPEN_TAG
        ];
    }

    /**
     * @inheritDoc
     */
    public function process(File $phpcsFile, $stackPtr)
    {
        $tokens = $phpcsFile->getTokens();

        $searchPtr = $stackPtr;
        $previousNewLine = false;
        while (--$searchPtr) {
            if ($tokens[$searchPtr]['code'] === T_WHITESPACE && $tokens[$searchPtr]['content'] === "\n") {
                // We have found the new line before the DocBlock
                if ($previousNewLine === true) {
                    return;
                } else {
                    $previousNewLine = true;
                }

                continue;
            }

            if ($tokens[$searchPtr]['code'] === T_SEMICOLON) {
                // We have encounter ; before new line
                $phpcsFile->addWarning(
                    'DocBlock and preceding statement must be separated with the new line',
                    $stackPtr,
                    'MissingNewLinePHPDoc'
                );
            }

            if ($tokens[$searchPtr]['code'] !== T_WHITESPACE) {
                // We have found something else, other than ;
                return;
            }
        }
    }
}


toolset/custom-standards/mcga/Sniffs/Commenting/MagentoCopyrightNoticeSniff.php
<?php
/**
 * Tools for static testing Magento code
 * Copyright (C) 2019 Stjepan Udovičić
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

namespace mcga\Sniffs\Commenting;

use PHP_CodeSniffer\Sniffs\Sniff;
use PHP_CodeSniffer\Files\File;

/**
 * Requires removal of original copyright notice when copy files over.
 */
class MagentoCopyrightNoticeSniff implements Sniff
{

    public function register()
    {
        return [
            T_DOC_COMMENT_STRING
        ];
    }


    /**
     * @inheritDoc
     */
    public function process(File $phpcsFile, $stackPtr)
    {
        $tokens = $phpcsFile->getTokens();
        $commentLine = strtolower($tokens[$stackPtr]['content']);

        if (strpos($commentLine, 'copyright') === false) {
            return;
        }

        if (strpos($commentLine, 'magento') === false) {
            return;
        }

        $phpcsFile->addWarning(
            'Original Magento copyright notice must be removed',
            $stackPtr,
            'InvalidCopyrightNotice'
        );
    }
}


toolset/custom-standards/mcga/Sniffs/Commenting/ConstantsPHPDocFormattingSniff.php
<?php
/**
 * Tools for static testing Magento code
 * Copyright (C) 2019 Stjepan Udovičić
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

namespace mcga\Sniffs\Commenting;

use PHP_CodeSniffer\Sniffs\Sniff;
use PHP_CodeSniffer\Files\File;

/**
 * Detects PHPDoc formatting for constants.
 */
class ConstantsPHPDocFormattingSniff implements Sniff
{
    /**
     * @inheritDoc
     */
    public function register()
    {
        return [
            T_CONST,
            T_STRING
        ];
    }

    /**
     * @inheritDoc
     */
    public function process(File $phpcsFile, $stackPtr)
    {
        $tokens = $phpcsFile->getTokens();

        if ($tokens[$stackPtr]['code'] != T_CONST
            && !($tokens[$stackPtr]['content'] == 'define' && $tokens[$stackPtr+1]['code'] == T_OPEN_PARENTHESIS)
        ) {
            return;
        }

        $constNamePtr = $phpcsFile->findNext(
            ($tokens[$stackPtr]['code'] === T_CONST) ? T_STRING : T_CONSTANT_ENCAPSED_STRING,
            $stackPtr + 1,
            null,
            false,
            null,
            true
        );
        $constName = strtolower(trim($tokens[$constNamePtr]['content'], " '\""));

        $commentStartPtr = $phpcsFile->findPrevious(T_DOC_COMMENT_OPEN_TAG, $stackPtr - 1, null, false, null, true);
        if ($commentStartPtr === false) {
            return;
        }

        $commentCloserPtr = $tokens[$commentStartPtr]['comment_closer'];
        for ($i = $commentStartPtr; $i <= $commentCloserPtr; $i++) {
            $token = $tokens[$i];

            // Not an interesting string
            if ($token['code'] !== T_DOC_COMMENT_STRING) {
                continue;
            }

            // Comment is the same as constant name
            $docComment = trim(strtolower($token['content']), ',.');
            if ($docComment === $constName) {
                continue;
            }

            // Comment is exactly the same as constant name
            $docComment = str_replace(' ', '_', $docComment);
            if ($docComment === $constName) {
                continue;
            }

            // We have found at lease one meaningful line in comment description
            return;
        }

        $phpcsFile->addWarning(
            'Constants must have short description if they add information beyond what the constant name supplies.',
            $stackPtr,
            'MissingConstantPHPDoc'
        );
    }
}
